File: src/232.cpp
class MyQueue {
    /*
    time defeat: 100%
    space defeat: 74.23%
    */
private:
    stack<int> mainStack;   //用来实现先入先出操作的主栈
    stack<int> cacheStack;  //用来缓存元素的缓存栈
public:
    /** Initialize your data structure here. */
    MyQueue() {
        //可以不做吧
    }
    
    /** Push element x to the back of queue. */
    void push(int x) {
        //只向主栈推入元素
        mainStack.push(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        int topNum = 0, tempNum = 0;
        //把主栈中的元素都转移到缓存栈去
        while(mainStack.size() != 1){
            topNum = mainStack.top();   //获得栈顶元素
            cacheStack.emplace(topNum); //压入缓存
            mainStack.pop();    //弹出
        }
        //然后记录当前的最前元素
        topNum = mainStack.top();
        mainStack.pop();
        //然后再从缓存栈压回去
        while(!cacheStack.empty()){
            tempNum = cacheStack.top();
            mainStack.emplace(tempNum);
            cacheStack.pop();
        }
        return topNum;
    }
    
    /** Get the front element. */
    int peek() {
        int topNum = 0, tempNum = 0;
        //把主栈中的元素都转移到缓存栈去
        while(mainStack.size() != 1){
            topNum = mainStack.top();   //获得栈顶元素
            cacheStack.emplace(topNum); //压入缓存
            mainStack.pop();    //弹出
        }
        //然后记录当前的最前元素
        topNum = mainStack.top();
        //然后再从缓存栈压回去
        while(!cacheStack.empty()){
            tempNum = cacheStack.top();
            mainStack.emplace(tempNum);
            cacheStack.pop();
        }
        return topNum;
    }
    
    /** Returns whether the queue is empty. */
    bool empty() {
        return mainStack.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: src/232.py
class MyQueue:
    #要用两个栈
    #一个主栈，一个在出队时的辅助栈
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.mainStack = list() #后进后出
        self.assistStack = list()


    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        self.mainStack.append(x)


    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        #弹出主栈元素到辅助栈中
        while len(self.mainStack):
            self.assistStack.append(self.mainStack[-1])
            #弹出元素
            self.mainStack.pop()
        result = self.assistStack[-1]   #获得栈顶
        self.assistStack.pop()
        while len(self.assistStack):
            self.mainStack.append(self.assistStack[-1])
            self.assistStack.pop()
        return result


    def peek(self) -> int:
        """
        Get the front element.
        """
        #弹出主栈元素到辅助栈中
        while len(self.mainStack):
            self.assistStack.append(self.mainStack[-1])
            #弹出元素
            self.mainStack.pop()
        result = self.assistStack[-1]   #获得栈顶
        while len(self.assistStack):
            self.mainStack.append(self.assistStack[-1])
            self.assistStack.pop()
        return result

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
        return len(self.mainStack) == 0



# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()
